docs/tsm/INDEXING_REQS.md

```diff
@@ -0,0 +1,59 @@
+# Overview
+
+This document describes the "high cardinality" problem as it relates to the in-memory index.  The current problems and the requirements needed for a solution to the problem are explained.
+
+## Problem Statement
+
+The database maintains an in-memory, inverted index that maps measurements with tags and values to series.  This was initially implemented as part of previous storage engines to facilitate faster query planning and meta queries.
+
+The in-memory aspect of the index present problems when series cardinality is large.  This can happen from writing to many distinct measurements or tags as well as using tag keys and values that are large in size.  For example, writing a Docker container ID create high cardinality as well as large tag values consuming large amounts of memory.  Storing counter as a tag can create many distinct series causing memory problems over time.
+
+Another issue that arises with high cardinality is slower startup times.  Startup times increase because the index needs to be reloaded from TSM and WAL files.  Each series key must be re-parsed from the TSM file to reconstruct the measurement and tag to value index.
+
+## Requirements
+
+### Functional
+
+1. The index must support finding all series by a measurement
+2. The index must support finding all series by a tag and value
+3. The index must support retrieving all tag keys
+4. The index must support retrieving all tag values by tag key
+5. The index must support the removal of measurements
+6. The index must support removal of series
+7. The index must support removal of tag keys and values
+8. The index must support finding all series by regex
+9. Updating the index for new series must not cause writes to block
+10. Queries and writes must not block each other
+11. The index must support point-in-time snapshots using the current shard backup and snapshotting mechanism (hard links)
+
+### Performance
+
+1. The index must be able to support 1B+ series without exhausting RAM
+2. Startup times should must not exceed 5 mins
```

Comment:
I agree and I think we should aim to have it lower.  I used 5 mins because with one restart in a year, that would be the amount of downtime allowed for five 9s.